### PR TITLE
Relative transform

### PR DESCRIPTION
**Pull request by SleepingSoul**

### CODE REVIEW

File: SleepEngine/Engine/Components/Transform2D.h [hunk 1]
```diff
@@ -8,6 +8,11 @@ class Transform2D : public Component
 {
 public:
     using Base = Component;
+    struct RelativeTransform
+    {
+        glm::vec2 Position{ 0.f, 0.f };
+        float Rotation{ 0.f };
+    };
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Components/Transform2D.h]: Why nest this. It is meant to be used outside of this class, sure they are releated, but not as much. Putting it in the slp namespace would clean code a lot.

File: SleepEngine/Engine/ObjectTree.cpp [hunk 2]
```diff
@@ -12,7 +12,7 @@ void ObjectTree::update(float dt)
     }
 }
 
-void ObjectTree::addToRoot(ObjectsContainer::value_type&& object)
+void ObjectTree::addToRoot(std::unique_ptr<Object>&& object)
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/ObjectTree.cpp]: yyeeah, i hated the unnesesarry level of abstraction.

File: SleepEngine/Engine/Components/HierarchyNode.cpp [hunk 3]
```diff
@@ -7,57 +7,75 @@ BEGIN_NAMESPACE_SLEEP
...
     Base::update(dt);
-    for(auto& childNodeObject : m_children)
+
+    for (auto& childNodeObject : m_children)
     {
         childNodeObject->update(dt);
     }
 }
+
+void HierarchyNode::setUpChild(Object& child) const
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Components/HierarchyNode.cpp]: this function is used only once, and has a poor name - too generic ,and i presume - purpose. You've basically just moved all code from addChild to here. It is also used only once. I think it hinders readability and elegance, i'd suggest you just remove it. As of now

File: SleepEngine/Engine/Components/HierarchyNode.h [hunk 4]
```diff
@@ -5,11 +5,13 @@ BEGIN_NAMESPACE_SLEEP
...
 {
-    using Base = Component;
 public:
+    using Base = Component;
+    using ChildrenContainer = std::vector<std::unique_ptr<Object>>;
+
     HierarchyNode();
 
-    GET_SET(Object*, getParent, setParent, m_object)
+    GET_SET(Object*, getParent, setParent, m_parent)
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Components/HierarchyNode.h]: Also, please rename the getParent/setParent form Component class to getObject/setObject. As of now this shadows that declaration, which is very bad. My bad, forgot to rename that myself.

File: SleepEngine/Engine/Components/HierarchyNode.cpp [hunk 5]
```diff
@@ -7,57 +7,75 @@ BEGIN_NAMESPACE_SLEEP
...
         LOG_AND_FAIL("parent is null");
         return;
     }
 
-    auto* parentNode = m_object->getComponent<HierarchyNode>();
-    parentNode->removeChild(m_object);
-    m_object = nullptr;
+    auto* const parentNode = m_parent->getComponent<HierarchyNode>();
+    parentNode->removeChild(m_parent);
+    m_parent = nullptr;
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Components/HierarchyNode.cpp]: Let's init the trivial members inline& its what we both were doing so far from what i could see

Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: But this is not an initialization

Comment (TarasMartynyuk) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: oh, meant to comment other line 
image

Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: I just don't like any initialization in header file if we have .cpp.
One thing to do it if it is a simple struct. But in class with interface, it looks like unnecessary information in the header.
In my opinion, it's pretty ok for Transform2DData, for example, but not for this class.

Comment (TarasMartynyuk) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: maybe, but it is easier to track that way. If you have more than 1 constructor
plus, you don't need default constructor just to init the variables

Comment (TarasMartynyuk) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: Plus, the members are implementation detail anyway.

Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: Sure. Fixed

File: SleepEngine/Engine/Components/HierarchyNode.h (hunk 4 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.h]: Yeah, thanks. Done

File: SleepEngine/Engine/Components/HierarchyNode.cpp (hunk 3 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: Ok, done

File: SleepEngine/Engine/Components/Transform2D.h (hunk 1 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/Components/Transform2D.h]: Moved to a different file. Let it be like this. In the future, I will do caching for global data and we will get rid of this struct.

**Pull request merged**